### ISSUE TITLE
[TextInput] Dynamic changing of secureTextEntry breaks font and doesn't change cursor position

### ISSUE THREAD

**Issue by aksonov** [facebook/react-native#5859]
I need to have show/hide password option for TextInput.
Dynamic changing of secureTextEntry breaks font and doesn't change cursor position. It probably happens because font for "*" is fixed and it is not changed back when securyTextEntry is false.

**Comment by kocyigityunus** [comment]
You are right. that's definitely a bug. But checkout my solution below. 

İf you add a space character and delete it again after the cursor will be at the right position. 

```javascript
/*jslint node: true */
'use strict';
import React, {
  Component,
  StyleSheet,
  View,
  TextInput,
  Switch
} from 'react-native';

class Example extends Component {
  constructor( props) {
    super(props);

    this.state = {
      isSecureTextEntry : true,
      password : '123456'
    };
  }

  render() {
    return (
      <View style={{
          flex : 1,
          backgroundColor : '#F5FCFF'
        }}>
        <Switch
          onValueChange={(value) =>{
            this.setState({isSecureTextEntry: value});
            this.setState({ password : this.state.password + ' ' });
            this.setState({ password : this.state.password.substring(0, this.state.password.length - 1)});
            }}
          style={{marginLeft: 20, marginTop : 50}}
          value={this.state.isSecureTextEntry}
          />
          <TextInput
            style={{ height: 40, borderColor: 'gray',
              borderWidth: 1, marginLeft : 20, marginTop : 50}}
            onChangeText={(text) => this.setState({password : text})}
            secureTextEntry = {this.state.isSecureTextEntry}
            value={this.state.password}
            />
      </View>
    );
  }
}

module.exports = Example;

```

**Comment by christopherdro** [comment]
@facebook-githb-bot bugfix

**Comment by grabbou** [comment]
This is going to be fixed by #6564 I believe.

**Comment by charpeni** [comment]
@grabbou You're right, #6564 has landed into 0.26.

**Comment by charpeni** [comment]
@facebook-github-bot close

**Comment by facebook-github-bot-8** [comment]
@charpeni tells me to close this issue. If you think it should still be opened let us know why.

**Comment by sunnylqm** [comment]
Still exists in 0.26.
![qq 20160603133122](https://cloud.githubusercontent.com/assets/615282/15769638/8955ebb2-298f-11e6-8e5c-fa29c3e3b8c7.jpg)
![qq 20160603133127](https://cloud.githubusercontent.com/assets/615282/15769643/91928c9a-298f-11e6-85d8-0d140ed34efd.jpg)

@grabbou @charpeni

**Comment by sunnylqm** [comment]
@facebook-github-bot reopen

**Comment by johnwonder** [comment]
Still exists in 0.27

**Comment by jevakallio** [comment]
Appears to still exist in iOS in 0.30, as reported in [this SO question](http://stackoverflow.com/questions/38707865/react-native-ios-textinput-switching-securetextentry-switches-font).

I also previously reported this with dupe https://github.com/facebook/react-native/issues/5710.

@charpeni could we consider reopening this?

**Comment by mmazzarolo** [comment]
Yup, still exists on 0.31-rc

**Comment by rkostrab** [comment]
I've written a [thread](http://stackoverflow.com/questions/38707865/react-native-ios-textinput-switching-securetextentry-switches-font?noredirect=1#comment64796566_38707865) about this.

**Comment by kacperpotega** [comment]
Still exists in 0.33.0.

I'm also experiencing an additional issue when updating secureTextEntry dynamically. When changing from false to true, the text passed into the onChangeText function is empty when additional text is typed.

**Comment by charpeni** [comment]
@facebook-github-bot reopen

**Comment by facebook-github-bot-4** [comment]
Okay, reopening this issue.

**Comment by ericnakagawa** [comment]
Hey folks, am closing this issue as we're at v40 now. If it's still persistent, reopen this issue and I'll find someone here at Facebook to investigate.

**Comment by diegorodriguesvieira** [comment]
This bug still happens on iOS version 9.6 (iPhone 4S)
![image](https://media.giphy.com/media/3o84TSWWRQWKMHceRy/giphy.gif)

**Comment by sunweiyang** [comment]
This is still happening on RN 0.50.4 -- any updates?

**Comment by grabbou** [comment]
I've reopened this issue. If you are thinking of contributing this fix, please let me know - I am more than happy to help you and give as many ideas as possible.

**Comment by Diego-Franco** [comment]
you still having the same problem on Android @grabbou when you enable/disable the "secureTextEntry" the cursor for TextInput is moving to the position 0 of the string.

**Comment by rimallo** [comment]
Still exists. 
- Adding ' ' space and removing removes last character because of 2 calls setState one after another.
- Giving blur() hides Keyboard

**Comment by jhnferraris** [comment]
I'm still experiencing the bug where cursor position resets when "secureText" prop changes

**Comment by hramos** [comment]
Tagging as a "Good first issue", please reach out to Grabbou if you are interested in working on a fix.

**Comment by grabbou** [comment]
Happy to help everyone who is interested in landing a fix for it :)

**Comment by jayporta** [comment]
I'd like to give it a shot.

**Comment by grabbou** [comment]
Go ahead. Happy to help.

**Comment by lucasbento** [comment]
@grabbou: can you provide a small summary of where to fix it? Is it related to the current selection of `TextInput` or the font?

**Comment by lucasbento** [comment]
@grabbou: opened PR #18587, would you mind taking a look? I'm not sure if that was the best place to implement it but it does fix the problem. :)

**Comment by jayporta** [comment]
Every time I tried to fork the repo in order to work on this issue, GitHub would just display a loading graphic. I can fork other repos without a problem, so it doesn’t seem to be just a GitHub problem. Am I missing something?

**Comment by rajat-np** [comment]
I forked this repo just now. There was no problem. @jayporta

**Comment by ButuzGOL** [comment]
Temporary workaround
```
this.inputRef.setNativeProps({
        selection: { start: 0, end: 0 },
      });
      this.inputRef.setNativeProps({
        selection: {
          start: passwordLength,
          end: passwordLength,
        },
      });
```

**Comment by sghosh968** [comment]
@ButuzGOL : I tried the temporary fix you suggested, but didn't seem to work for me.
I am invoking **setNativeProps** in **setState** callback, where I update state variable used to toggle **secureTextEntry**.
`
this.setState(prevState => ({
  hidePassword: prevState.hidePassword
}), () => {
  const currentPasswordLength = this.state.password.length;
  const passwordInputRef = this.inputs.password;
  if (passwordInputRef) {
    passwordInputRef.setNativeProps({
      selection: {
        start: currentPasswordLength,
        end: currentPasswordLength
      }
    });
  }
});
`
I guess I am putting **setNativeProps** in a wrong place, not sure though.
Any help on this is greatly appreciated.

**Comment by lucasboleli** [comment]
I'm having the same problem:

`.setNativeProps({ selection: { start: currentPasswordLength, end: currentPasswordLength } });`

this is not changing the cursor position.

**Comment by zhonggithub** [comment]
I  having same problem .
the react native version is 0.55.4

**Comment by sghosh968** [comment]
This is for those who were facing the following issue.
On Android when dynamically updating **secureTextEntry** for a **TextInput** cursor moves to zero every time.
It has been fixed in **v0.54.0**.
**Link to commit :** https://github.com/facebook/react-native/commit/09b43e4

**Comment by hramos** [comment]
You can find a full repro of this here: https://snack.expo.io/Bk1CsdRE7
Credit to @kocyigityunus for the original repro at https://github.com/facebook/react-native/issues/5859#issuecomment-182584283 above.

Repro steps:

1. Open app.
2. Tap switch to off position, observe "12345" instead of "123456" in text field as expected.
3. Tap switch back to on, then off position. Observe "1234" this time around. Rinse, repeat

**Comment by andrejcesen** [comment]
@sghosh968 I am still experiencing this issue in RN **v0.56.0**. Could you post an example where this is working as expected?

**Comment by sghosh968** [comment]
Sorry for posting without checking first looking at the gif it seemed the issue was fixed, should have checked first. @andrejcesen, I recently upgraded an app from  **v0.52** to **0.56** and found the issue still persists.

**Comment by lucasbento** [comment]
PR https://github.com/facebook/react-native/pull/18587 fixes this, I just need a hand from a maintainer/someone who knows more of Objective-C to reiterate based on @janicduplessis requests.

**Comment by AnitaSinghLoved** [comment]
`DIRTY HACK`
just changing the fontSize will do the trick:
```fontSize: (this.state.showPassword) ? 24 : 23```

**Comment by tuannguyenhoangit** [comment]
> This is for those who were facing the following issue.
> On Android when dynamically updating **secureTextEntry** for a **TextInput** cursor moves to position zero every time.
> It has been fixed in **v0.54.0**.
> **Link to commit :** [09b43e4](https://github.com/facebook/react-native/commit/09b43e4)

0.55.4 still get issue

**Comment by bhaskarvgurram** [comment]
I have the same issue on 0.57.5

**Comment by bhaskarvgurram** [comment]
> You can find a full repro of this here: https://snack.expo.io/Bk1CsdRE7
> Credit to @kocyigityunus for the original repro at [#5859 (comment)](https://github.com/facebook/react-native/issues/5859#issuecomment-182584283) above.
> 
> Repro steps:
> 
> 1. Open app.
> 2. Tap switch to off position, observe "12345" instead of "123456" in text field as expected.
> 3. Tap switch back to on, then off position. Observe "1234" this time around. Rinse, repeat

The issue is still not resolved. You can check it here : https://snack.expo.io/Bk1CsdRE7

**Comment by CatapultJesse** [comment]
> You can find a full repro of this here: https://snack.expo.io/Bk1CsdRE7
> Credit to @kocyigityunus for the original repro at [#5859 (comment)](https://github.com/facebook/react-native/issues/5859#issuecomment-182584283) above.
> 
> Repro steps:
> 
>     1. Open app.
> 
>     2. Tap switch to off position, observe "12345" instead of "123456" in text field as expected.
> 
>     3. Tap switch back to on, then off position. Observe "1234" this time around. Rinse, repeat

i think this is because you're not using the setState callback, should be:
```
this.setState({ password : this.state.password + ' ' }, () => {
              this.setState({ password : this.state.password.substring(0, this.state.password.length - 1)});
            });
```
Can't get it to reproduce on android s8

**Comment by ericlewis** [comment]
Fixed: #23524.
